### ISSUE TITLE
After firmware update, error message with blue screen report that firmware signature verification failed..but system works perfect.

### ISSUE THREAD

**Issue by sateuwdie** [Dasharo/dasharo-issues#556]
**Dasharo version**
1.7.0

**Dasharo variant**
 Dasharo (coreboot+UEFI) v1.7.0

**Question text**
Today I have upgraded the firmware via ipxe from dasharo tools suite, version: 1.7.0
The update apply but the system poweroff immedately during update (of course was connected to AC source)
i press the power button and system come up..with a blue screen tell me firmware was not updated correctly.
The system boot, all works, no error reported, I can enter in coreboot menu, etc.
Can I stay safe? Thanks
The system is NS5x_NS7xPU

**Comment by wessel-novacustom** [comment]
Thank you for your report!

The immediate power off behaviour is normal as the EC firmware needed an update as well, this cannot be prevented.

We asked our team to stop the updates from being public, as there are some critical issues with the latest versions.

Nevertheless, I believe it can be ignored safely, as this is probably a result of the entire BIOS region being flashed. I also believe that the blue screen you saw was a message that the firmware could not verify its integrity because of a newer EDK II version. This hypothesis hasn't been verified by the developers yet, though.

Please contact us directly in case you need instructions to downgrade to v1.5.0.

If you do not experience any critical issues with v1.7.0, you can safely continue to use it. A hotfix release (v1.7.1) will be published very soon.

**Comment by sateuwdie** [comment]
Thanks for answer.
I reply to all

a) No critical issues: except the keyboard backlight remain on and I cannot know how to turn off (someone know? I have tried `echo 0 > /sys/class/leds/system76_acpi\:\:kbd_backlight/brightness` but is still on.

b)Yes, the blue screen image was this

https://user-images.githubusercontent.com/19353650/279761853-00d02d46-3d1a-4a44-a964-8c9b42dbd379.jpg

c)the previous firmware version was 1.6.0

**Comment by wessel-novacustom** [comment]
I got the same message when upgrading from v1.6.0 to v1.7.1.

**Comment by macpijan** [comment]
@wessel-novacustom Did you have the same exact 0x5F error as on the screen above?

**Comment by macpijan** [comment]
Some ideas from discussions so far:
- we inform in this message that we have booted from the recovery partition
- we might display this message if **any** error code appears, but maybe not all of them means the recovery request has happened?
- we need to have a closer look on the implementation details here.

**Comment by pietrushnic** [comment]
@macpijan related issues

- https://github.com/Dasharo/dasharo-issues/issues/489
- https://github.com/Dasharo/dasharo-issues/issues/442
- https://github.com/Dasharo/dasharo-issues/issues/269

There are a number of issues regarding this feature, IMHO, which means there may be some design flaws. What bothers me is the fact that we easily display messages to scare users, but we do not provide enough guidance on the real meaning of the problem and recovery mechanics.

So, we implemented a protection mechanism with poor detection quality (many false positives) and no recovery. Every design of security features should treat all stages of protection, detection, and recovery equally seriously.

**Comment by wessel-novacustom** [comment]
I just reproduced this issue on the NS51PU-i7 after flashing v1.7.1

**Comment by mkopec** [comment]
I believe if we did updates *only* for the RW_SECTION_A/B regions, the issue would no longer apply. Updating the RO partition will cause problems such as these. It also makes sense from the security perspective and would allow us to use the SPI chip's flash protection.

This is a big advantage that Vboot gives us and we're not making use of it.

**Comment by wessel-novacustom** [comment]
@macpijan I think we should try that tomorrow.

**Comment by macpijan** [comment]
> detection quality (many false positives) and no recovery

This is not really proven yet. Maybe we do in fact boot from recovery in these cases, so the message would be true. We need to have a closer look at the case when reproduction happens.

**Comment by pietrushnic** [comment]
> This is not really proven yet. Maybe we do in fact boot from recovery in these cases, so the message would be true. We need to have a closer look at the case when reproduction happens.

@macpijan Yes, the technical root cause is not yet proven, and I can't comment on the reason in the code because I don't know this implementation so well. My comment is purely how it seems from a UX perspective. We need better tests of this feature.

**Comment by miczyg1** [comment]
@macpijan upgrading from v1.7.0 to v.1.7.1 does not reproduce this issue. How to get DTS image which upgrades from v1.5.0 to v1.7.0? My guess is the DTS v1.2.10 but the files are nowhere to be found. I would like to use the same procedure when reproducing it, especially that the error occurred only once for unknown reason (the hash space lock was in v1.5.0 too).

**Comment by macpijan** [comment]
@miczyg1 Wessel has been able to reproduce it going from 1.6.0 to 1.7.1 with the DTS available in iPXE.

**Comment by miczyg1** [comment]
> This error has happened multiple times already, for Marek, Wessel, and various end users.

I meant the error occurs only once after flashing.

So all it takes is to flash v1.6.0 back and update from DTS again

**Comment by miczyg1** [comment]
v1.6.0 -> v.1.7.1 still no reproduction.

**Comment by miczyg1** [comment]
Found possible culprit:
https://github.com/Dasharo/coreboot/blob/dasharo/src/mainboard/clevo/adl-p/cmos.layout#L22
perfectly overlaps with https://github.com/Dasharo/coreboot/blob/dasharo/src/soc/intel/common/basecode/ramtop/ramtop.c#L35

Having such luck is beyond coincidence...

**Comment by miczyg1** [comment]
https://github.com/Dasharo/coreboot/pull/434

**Comment by wessel-novacustom** [comment]
I will need to test this on a brand new device that has never had Dasharo installed before. For example, by installing v1.6.0 on NS5x 12th Gen, install the operating system and use it normally. Then I will install the test binary with the command `flashrom -p internal --ifd -i bios -w /path/to/coreboot.rom` to see if I get the message again.

**Comment by pietrushnic** [comment]
Another instance of this problem was reported on `Dasharo - Support` channel. Fixing it is one thing, but correctly communicating such a scary message may pop up is another. IT would be great if we can address both, because this bug really damage Novacustom and Dasharo reputation.

**Comment by BeataZdunczyk** [comment]
Your insights are highly appreciated, @sycam0r-e, thank you for your contributions!

>  Let me know if and where I should comment on that. :)

Please consider creating issues on our [Dasharo Issues](https://github.com/Dasharo/dasharo-issues/issues) page.

You can conveniently create new issues using this link: [Create a New Issue](https://github.com/Dasharo/dasharo-issues/issues/new/choose). Providing detailed context and information, including screenshots or photos, will help us address and resolve the issues effectively.

Feel free to join our Matrix community: https://matrix.to/#/#dasharo:matrix.org to share your thoughts.

**Comment by sycam0r-e** [comment]
Allow me to direct your attention to one of the more important parts of my earlier message that likely just got lost:

>As far as I understand your current thinking is that the warning message being shown is erroneous and that users have actually nothing to worry about. However, this is not stated clearly in this issue. **_Make a separate comment in this issue detailing under which conditions (e.g. if the iPXE boot menu has been used) users do not have to worry._**

Emphasis added.

**Comment by macpijan** [comment]
Issue has been reopened, as it seems the problem has not been solved, and it still appears in the following scenario:

1. Get a new device with proprietary firmware
2. Flash Dasharo 1.6.0 (or older)
3. Update to Dasharo 1.7.2
4. The popup appears on the first boot.
5. This problem cannot be reproduced on the same hardware unit again thus far, making it very tricky to diagnose

In: https://github.com/Dasharo/dasharo-issues/issues/556#issuecomment-1809940586 there was a potential fix, but it seems that the problem still appears on the new hardware units.

**Comment by wessel-novacustom** [comment]
@macpijan Exactly.

As we have new hardware all the time, we can do testing for this issue.

**Comment by marmarek** [comment]
I believe I hit this issue on MSI Z690 too. Updated from 1.1.1 to 1.1.3 using `flashrom -p internal -w [path] --fmap -i RW_SECTION_A -i RW_SECTION_B`, then typed "reboot" and got message like this:
![Dasharo-MSI-1 1 1-1 1 3-update](https://github.com/Dasharo/dasharo-issues/assets/726704/b7b1dfea-c641-4f5a-98ab-c8577909f911)

After pressing ENTER, I immediately entered setup, and was greeted with 1.1.1 version there.

**Comment by mkopec** [comment]
@marmarek I believe the command for updating to v1.1.3 should be `flashrom -p internal --ifd -i bios` according to the DTS scripts: https://github.com/Dasharo/meta-dts/blob/017ec32aab3f4481c3a904d507212fcf73570cc7/meta-dts-distro/recipes-dts/dts/dts/dts-functions.sh#L323

**Comment by marmarek** [comment]
hmm, maybe better describe it on https://docs.dasharo.com/unified/msi/firmware-update/ then? the section "Version v1.1.0 or newer" says to do just RW_SECTION_{A,B} (there is a note about 1.1.2 too, but it's unclear...)

**Comment by mkopec** [comment]
Yes, the docs appear to be outdated, see my edit above. Need to update the docs so that the entire BIOS region is flashed.

**Comment by marmarek** [comment]
ok, after updating entire BIOS region, correct version is running, but still got the message after update.

**Comment by mkopec** [comment]
The error indicates that Vboot could not save the hash of recovery MRC cache to TPM NVRAM. Since this error appears once per machine lifetime (after switching from Insyde), it may mean that we need to do a reset of the TPM during initial deployment. For upcoming releases we may work around this issue by turning on secdata mocking.

@miczyg1 wdyt?

**Comment by mkopec** [comment]
We've sent binaries for testing to @wessel-novacustom , please let us know when you're able to confirm that the issue is fixed now.

**Comment by wessel-novacustom** [comment]
> We've sent binaries for testing to @wessel-novacustom , please let us know when you're able to confirm that the issue is fixed now.

I'll do tomorrow at latest.

**Comment by wessel-novacustom** [comment]
I have tested the binary on an NV41PZ running Dasharo v1.6.0 (never been to v1.7.x on that unit before) with the following DTS commands:

wget http://192.168.1.75/coreboot.rom (this is from a local server of course)
sha256sum coreboot.rom says: ec05de09b986b45fb0d1e77262777c95a5a4f7f1ea58952dcbab30c0daf09bb6  coreboot.rom
I have flashed the binary with the following command:
flashrom -p internal --ifd -i bios -w coreboot.rom && systemctl poweroff
After a few minutes, the laptop has shut down.

I powered it on and the message did not appear as it usually does after an update.

I have tried this on multiple devices.

I have even updated manually from v1.6.0 to the original v1.7.2 version on an NV41-ADL that again never had Dasharo v1.7.x before. The pop up showed up at the first boot, which is not the case for the test binary.

So yes, we have the sweet spot, it seems.

**Comment by mkopec** [comment]
Closing, as the issue is confirmed to be fixed. I also posted some info in a related issue: https://github.com/Dasharo/dasharo-issues/issues/689#issuecomment-1981496430

**Comment by baptx** [comment]
@wessel-novacustom @mkopec It looks like the issue is not fixed yet because today I updated the firmware from version 1.6.0 to 1.7.2 on my NV41 Series laptop and the blue warning appeared once.

**Comment by wessel-novacustom** [comment]
> @wessel-novacustom @mkopec It looks like the issue is not fixed yet because today I updated the firmware from version 1.6.0 to 1.7.2 on my NV41 Series laptop and the blue warning appeared once.

It has been fixed for new releases, but there hasn't been any new stable version released yet.

Maybe @mkopec or @macpijan can confirm.
